Provide PHP 8.1 support

weierophinney opened the pull request:
This patch provides PHP 8.1 support for laminas-soap via the following changes:

- It adds `~8.1.0` to the list of supported PHP versions
- It **drops support for PHP 7.3**
- It changes how the package replaces zend-soap:
  - It renames the "replace" section to "conflict"
  - It changes the "zend-soap" entry in that section to have the constraint "*"
  - It removes the dependency on laminas-zendframework-bridge
- It adds `#[ReturnTypeWillChange]` attributes to the `Client` implementation where it overrides internal methods.
- It updates test assets to follow internal signatures.
- It adds Laminas CI configuration and pre-run scripts to allow testing against PHP 8.1.

weierophinney committed:
feat: prepare for PHP 8.1 support

- Add `~8.1.0` to list of supported PHP versions
- Bump dependencies to versions known to work with 8.1
- Change how the package replaces zend-server
  - Rename the "replace" section to "conflict"
  - Change the constraint of "zend-server" in that section to "*"
  - Remove the dependency on laminas-zendframework-bridge

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: ensure code runs on PHP 8.1

- Adds `#[ReturnTypeWillChange]` attribute to Client::__doRequest
- Updates test asset to follow SoapClient signature

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: Ensure CI pipeline can run correctly on 8.1

- Ignore platform PHP requirement on PHP 8.1
- For "lowest" builds, ensure we are on a laminas-code version that works with 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: switch from using typehints to #[ReturnTypeWillChange] attribute

Guarantees compatibility with PHP 7.3 and 7.4.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on .laminas-ci/pre-run.sh:
```diff
@@ -0,0 +1,27 @@
+fi
+
+cd "${WORKSPACE}"
+composer require "laminas/laminas-code:^4.4" --ignore-platform-req=php
```
Hmm, why is laminas-code needed?

froschdesign replied:
laminas-server uses laminas-code:

https://github.com/laminas/laminas-soap/blob/f4cff0772a8fc92ad1f841c0c8ab0b0f269c7dbc/composer.json#L21-L29

This pull request contains also an update for this:

```json
"laminas/laminas-server": "^2.11",
```

But I don't know why version 4.4 is explicitly installed in the pre-run script.

Ocramius replied:
But why is `laminas-code` not in `composer.json` then? :|

froschdesign replied:
Maybe wrong version, see in laminas-server: :confused:

```json
"require": {
    "php": "^7.3 || ~8.0.0 || ~8.1.0",
    "laminas/laminas-code": "^3.5.1 || ^4.0.0",
    "laminas/laminas-stdlib": "^3.3.1"
},
```

https://github.com/laminas/laminas-server/blob/3.0.x/composer.json#L21-L25

Ocramius replied:
We could declare a `"conflict": {"laminas/code": "<4.4"}` here instead, which is a bit healthier...

weierophinney replied:
Here's the situation:

laminas-server's latest release works against PHP 7.3 - 8.1. It depends on laminas-code 3.5 or 4.0. Since it's a _direct_ requirement, when Composer is resolving while running on PHP 8.1, it will pull a version that works with 8.1 even when the `--prefer-lowest` flag is used.

However, since laminas-code is a sub-dependency for laminas-soap (which depends on laminas-server), those same flags end up pulling the lowest supported version, in this case the 3.5.x releases. Those do NOT work with PHP 8.1, so I end up with deprecation notices.

If I were to do a "conflict" rule as you suggest, I'd also need to bump the minimum supported PHP version to 7.4 for this release, which feels off, as laminas-server doesn't have that same stipulation. So the pre-run hack was the workaround I used to ensure we get laminas-code >= 4.4.0 when running via lowest dependencies.

I can certainly bump the minimum supported PHP version here; I was mainly trying to keep parity with what laminas-serve supports.

Ocramius replied:
> If I were to do a "conflict" rule as you suggest, I'd also need to bump the minimum supported PHP version to 7.4 for this release, which feels off, as laminas-server doesn't have that same stipulation.

It wouldn't be a suggestion: it would be a requirement.

Also: YOLO. Ain't nobody got time for PHP 7.3.

weierophinney committed:
feat: remove support for PHP versions prior to 7.4

Allows simplifying CI, and ensuring we don't get false positive runs.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: conflict with laminas-code < 4.4

This ensures we get a laminas-code version that works with PHP 8.1.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update to laminas-coding-standard v2 series

First series to work with PHP 7.4.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: apply laminas-coding-standard 2.2 ruleset

Mostly ensuring parameters, return values, and properties have typehinting via either docblocks or explicitly (test and non-public elements only).

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove file-level docblocks

Removes the remaining `@see` annotation and file-level docblock from all files.

Replaces `@codingStandardsIgnoreFile` with `phpcs:disable`.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: simplify test bootstrap

No need for legacy bootstrap file for testing.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@Ocramius Okay, finished it up. The bump to 7.4 meant I needed to bump the CS version from the v1 series to the v2 series, which meant a lot more changes overall; I tried to do granular commits, with one being just the dep bump, the next applying autofixes and correcting stuff phpcs couldn't autofix, the next taking care of the remaining file-level docblocks, and finally "fixing" the PHPUnit bootstrap.

Ocramius reviewed (approved):
Didn't find anything dangerous in the diff (mostly looking for `==` to `===` changes, which are always a bit risky).

:ship: 

Thanks @weierophinney!
